### ISSUE TITLE
The usage of `action_spaces = gym.spaces.MultiDiscrete` in DQN policy, in dqn.py

### ISSUE THREAD

**Issue by Chrisa142857** [thu-ml/tianshou#236]
Hello!

I'm using `gym.spaces.MultiDiscrete` as my `action_spaces` in my custom env which has one agent and needs a multiple dimension action. A list of chooses YES/NO is as my input action, e.g. `[0 1 0 1 1 0]`.

However, the random action in the pre-collect stage is fixed as a single action, e.g. `[6 6 6 6 6 6 6 6]`, with the same shape as the neural network output using `argmax()` in `dqn.py`:
```python
        if hasattr(obs, "mask"):
            # some of actions are masked, they cannot be selected
            q_: np.ndarray = to_numpy(q)
            q_[~obs.mask] = -np.inf
            act = q_.argmax(axis=1)
        # add eps to act in training or testing phase
        if not self.updating and not np.isclose(self.eps, 0.0):
            for i in range(len(q)):
                if np.random.rand() < self.eps:
                    q_ = np.random.rand(*q[i].shape)
                    if hasattr(obs, "mask"):
                        q_[~obs.mask[i]] = -np.inf
                    act[i] = q_.argmax()
```
Hence I replace `q_.argmax()` with `to_numpy(torch.from_numpy(q_).max(dim=0)[1])` 
and `q_.argmax(axis=1)` with `act = to_numpy(torch.from_numpy(q_).max(dim=1)[1])`.

And there is my issue: Is that **allowed** in tianshou?

In Chinese: 
同学你好！我在使用`gym.spaces.MultiDiscrete`的时候发现tianshou的DQN policy会生成不对应的action，请问我的上述更改是否会对产生整个policy产生其他未知的影响？感谢同学~

**Comment by Trinkle23897** [comment]
I'm not quite sure what you mean. Could you please give me some extra explanations? Since `[0,1,0,1,1,0]` is MultiDiscrete action, I don't figure out why there is an action `[6,6,6,6,6,6,6,6]`.

In fact, the q value's space is [num_state, num_action]. If your action is 6 Y/N, you can unroll the action to 2^6=64, which means 0-63 and each number indicates a combination of 6 Y/N. Otherwise, you must change the current code to something like 6 q functions: `real_q = q0[s, a0] * q1[s, a1] * ... * q5[s, a5]`, where `a=[a0, a1, ..., a5]`.

**Comment by Chrisa142857** [comment]
> I'm not quite sure what you mean. Could you please give me some extra explanations? Since `[0,1,0,1,1,0]` is MultiDiscrete action, I don't figure out why there is an action `[6,6,6,6,6,6,6,6]`.
>
> In fact, the q value's space is [num_state, num_action]. If your action is 6 Y/N, you can unroll the action to 2^6=64, which means 0-63 and each number indicates a combination of 6 Y/N. Otherwise, you must change the current code to something like 6 q functions: `real_q = q0[s, a0] * q1[s, a1] * ... * q5[s, a5]`, where `a=[a0, a1, ..., a5]`.

My `action_spaces` setting code is
```python
action_spaces = MultiDiscrete([2, 2, 2, 2, 2, 2])
action = MultiDiscrete.sample()
> action = [0, 1, 1, 0, 1, 0]
```
And then I want the DQN's q value's space could be `[num_state, (2, num_action)]`, non-activated output in DNN, or `[num_state, num_action]` with sigmoid activation. So that the multiple dimension `action` could be working.

In `dqn.py`, the `action` from DNN is considered as the non-activated situation,

```python
act: np.ndarray = to_numpy(q.max(dim=1)[1])
```
which is worked with the action shape `(2, num_action)`. However, `act[i] = q_.argmax()` in random action sampling isn't working. Is it only consider the one dimension of q value? :)

**Comment by Trinkle23897** [comment]
> Is it only consider the one dimension of q value?

Sure

**Comment by danagi** [comment]
Just replace `act[i] = q_.argmax()` to `act[i]=q_.argmax(axis=0)` you'll find it works. But then you'll also find it's broken in function `learn`. That's because DQN doesn't support MultiDiscrete directly. The solution is to transform MultiDiscrete to Discrete. In your case, transform (2, num_action) to (2^num_action) and then transform it back in `env.step` which can be achieved by using custom `gym.wrapper`.
